internal/slicesx/heap.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slicesx
+
+import "cmp"
+
+// Heap is a port of Go's [container/heap] package to use generics instead of
+// interface calls. Instead, entries consist of an ordered key and an arbitrary
+// value.ß
+//
+// A zero heap is empty and ready to use.
+type Heap[K cmp.Ordered, V any] struct {
+	keys []K
+	vals []V
+}
+
+// Len returns the number of elements in the heap.
+func (h *Heap[K, V]) Len() int {
+	return len(h.keys)
+}
+
+// Push pushes the element x onto the heap.
+func (h *Heap[K, V]) Push(k K, v V) {
+	h.push(k, v)
+	h.up(h.Len() - 1)
+}
+
+// Pop removes and returns the entry with the least key from the heap.
+//
+// Pop is equivalent to [Heap.Remove](0).
+func (h *Heap[K, V]) Pop() (K, V) {
+	n := h.Len() - 1
+	h.swap(0, n)
+	h.down(0, n)
+	return h.pop()
+}
+
+// Remove removes and returns the entry at index i from the heap.
+func (h *Heap[K, V]) Remove(i int) (K, V) {
+	n := h.Len() - 1
+	if n != i {
+		h.swap(i, n)
+		if !h.down(i, n) {
+			h.up(i)
+		}
+	}
+	return h.pop()
+}
+
+// Fix re-establishes the heap invariant after the key of the nth entry changes.
+func (h *Heap[K, V]) Fix(n int) {
+	if !h.down(n, h.Len()) {
+		h.up(n)
+	}
+}
```

Comment:
The slices inside the heap are not exported. So aside from `Len()`, there is no information about an element's index. So these operations aren't particularly useful.